consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/SchedulableMiningCoordinator.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.consensus.common.bft.blockcreation;
+
+import static org.apache.logging.log4j.LogManager.getLogger;
+
+import org.hyperledger.besu.consensus.common.bft.BftForksSchedule;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.datatypes.Wei;
+import org.hyperledger.besu.ethereum.blockcreation.MiningCoordinator;
+import org.hyperledger.besu.ethereum.chain.BlockAddedEvent;
+import org.hyperledger.besu.ethereum.chain.BlockAddedObserver;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.Transaction;
+
+import java.util.List;
+import java.util.Optional;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+
+public class SchedulableMiningCoordinator implements MiningCoordinator, BlockAddedObserver {
+
+  private static final Logger LOG = getLogger();
+
+  private final BftForksSchedule<MiningCoordinator> miningCoordinatorSchedule;
+  private final Blockchain blockchain;
+  private MiningCoordinator activeMiningCoordinator;
+  private long blockAddedObserverId;
+
+  public SchedulableMiningCoordinator(
+      final BftForksSchedule<MiningCoordinator> miningCoordinatorSchedule,
+      final Blockchain blockchain) {
+    this.miningCoordinatorSchedule = miningCoordinatorSchedule;
+    this.blockchain = blockchain;
+    this.activeMiningCoordinator =
+        this.miningCoordinatorSchedule.getFork(blockchain.getChainHeadBlockNumber()).getValue();
+  }
+
+  @Override
+  public void start() {
+    blockAddedObserverId = blockchain.observeBlockAdded(this);
+    activeMiningCoordinator.start();
+    if (activeMiningCoordinator instanceof BftMiningCoordinator) {
```

Comment:
Will have debug through to understand this I think. But I don't see why the MiningCoordinator stop method that calls `blockchain.removeObserver` isn't sufficient. As we always call stop on the current miningCoordinator before calling start on the new one.